src/ctia/shutdown.clj

```diff
@@ -0,0 +1,22 @@
+(ns ctia.shutdown
+  (:require [clojure.tools.logging :as log]))
+
+(defonce shutdown-hooks (atom {:registered? false :hooks {}}))
+
+(defn register-hook! [key f]
+  (swap! shutdown-hooks assoc-in [:hooks key] f))
+
+(defn shutdown-ctia! []
+  (doseq [[name hook] (:hooks @shutdown-hooks)]
+    (try
+      (hook)
+      (catch Exception excep
+        (log/error excep (format "Shutdown hook %s failed" name)))))
+  (swap! shutdown-hooks assoc :hooks {}))
+
+(defn init! []
+  (when-not (:registered? @shutdown-hooks)
+    (.addShutdownHook
+     (Runtime/getRuntime)
+     (Thread. shutdown-ctia!))
+    (swap! shutdown-hooks assoc :registered? true)))
```

Comment:
That's a good point, but we really don't need to support concurrency.  If we did, I'd want to use an agent (for side-effects) to set the hook.  Its not needed, but I'll change it, just to be pedantic.  :)